### PR TITLE
Add flag to set namespace in TF

### PR DESCRIPTION
**Pull request by RPS98**
---
 
## Basic Info
 
| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | - |
| Primary OS tested on | Ubuntu |
| Drone platform tested on | DJI M300 RTK |
 
---
 
## Description of contribution in a few bullet points
 
* Add the flag `add_namespace_to_tf` in `psdk_params.yaml` that allows adding node namespace to TF frames.

## Motivation and Context

This feature is required when working with several drones in a swarm and there's a need to identify the TF frames of each drone. This can be achieved either by using different `psdk_params.yaml` files for each namespace or by creating a custom launcher that declares these parameters. But I think this solution is more user-friendly.

### CODE REVIEW

**Comment by biancabnd** [comment]
Hi @RPS98, 
Thank you for all your contributions. 
Regarding this PR, can you explain me a bit why this new parameter is needed? Technically, you can just edit the frame id name directly. Do you have multiple psdk applications under the same ROS domain ID and you need the TF information of all copters to be shared?

**Comment by RPS98** [comment]
Hi @biancabnd,

I have a swarm of several DJI drones, and to process the data from all of them, such as their position or visualization in Rviz, I need to have the TF tree of all of them connected. The problem is that if multiple different frames have the same name, the tree breaks.

The solutions are:
- That flag that allows adding TFs in a specific drone namespace.
- A different configuration file for each drone that changes the name of the frames.

I believe the second option is more complex, as it complicates automation by requiring a separate file for each existing namespace. Nonetheless, this option would be functional if frames `h20_zoom_optical_link` and `h20_wide_optical_link` were in the the configuration file, so they would need to be added.

Which option do you think is the best?

**Comment by biancabnd** [comment]
> Hi @biancabnd,
> 
> I have a swarm of several DJI drones, and to process the data from all of them, such as their position or visualization in Rviz, I need to have the TF tree of all of them connected. The problem is that if multiple different frames have the same name, the tree breaks.
> 
> The solutions are:
> 
> * That flag that allows adding TFs in a specific drone namespace.
> * A different configuration file for each drone that changes the name of the frames.
> 
> I believe the second option is more complex, as it complicates automation by requiring a separate file for each existing namespace. Nonetheless, this option would be functional if frames `h20_zoom_optical_link` and `h20_wide_optical_link` were in the the configuration file, so they would need to be added.
> 
> Which option do you think is the best?

Hi @RPS98, 
I understand what you are saying, but as you know there is no concept of namespace within the TF tree. All frame names are just strings which you could differentiate among by adding the namespace in front but that is just one particular strategy.  I am trying to see which is the best way to do this while keeping it as general and useful as possible for all users. For example, we do not use `namespace/frame_id` to differentiate among frames but use `namespace_frame_id`. 

Having said that, I think we can leave what you are suggesting in this PR but removing the "/", this could go in each frame name. So we would just append to each frame the namespace without any separation. How does that sound to you?

File: psdk_wrapper/cfg/psdk_params.yaml [hunk 1]
```diff
@@ -14,6 +14,7 @@
       map_frame: "psdk_map_enu"
       gimbal_frame: "psdk_gimbal_link"
       camera_frame: "psdk_camera_link"
+      add_namespace_to_tf: false
```
Comment (biancabnd) [inline psdk_wrapper/cfg/psdk_params.yaml:17]: Please add this new parameter in the documentation.

File: psdk_wrapper/src/psdk_wrapper.cpp [hunk 2]
```diff
@@ -412,6 +413,20 @@ PSDKWrapper::load_parameters()
...
   }
+  if (!get_parameter("add_namespace_to_tf", params_.add_namespace_to_tf))
+  {
+    RCLCPP_WARN(get_logger(),
+                "add_namespace_to_tf param not defined, using default one: %s",
+                params_.add_namespace_to_tf ? "true" : "false");
+  }
+  if (params_.add_namespace_to_tf)
+  {
+    params_.imu_frame = std::string(get_namespace()) + "/" + params_.imu_frame;
```
Comment (biancabnd) [inline psdk_wrapper/src/psdk_wrapper.cpp:424]: remove the "/" and just append the namespace. the separator can be in the frame id string

File: psdk_wrapper/cfg/psdk_params.yaml (hunk 1 above)
Comment (biancabnd) [inline psdk_wrapper/cfg/psdk_params.yaml:17]: Add this new parameter in the documentation

**Review by biancabnd — COMMENTED** [review]
Overall nice addition! :confetti_ball:

**Comment by RPS98** [comment]
The use of '/' for TF names separation comes from ROS 1, where the tf_prefix existed. In ROS 2, this utility has been removed ([ROS Wiki](https://wiki.ros.org/tf2/Migration#Removal_of_support_for_tf_prefix)). Therefore, there is currently no standard way to separate TF names ([ROS answers 1](https://answers.ros.org/question/405822/multi-robot-tf-with-namespaces-for-tf-or-with-frame_prefix-what-is-the-right-way-to-do-it-ros2/) and [ROS answers 2](https://robotics.stackexchange.com/questions/100288/a-definitive-guide-to-launching-multi-robot-operations-in-ros2)). My proposal was to continue using the ROS 1 standard, as there are guidelines that establish it for ROS 2 ([The Construct](https://www.theconstruct.ai/ros2-qa-how-to-setup-identical-robots-for-multi-robot-navigation-236/)). What do you think?

I have modified the code, making the frame change through a class method ([`generate_tf_name`](https://github.com/RPS98/psdk_ros2/blob/add_namespace_to_tf/psdk_wrapper/include/psdk_wrapper/psdk_wrapper.hpp#L2054-L2070)), so that making changes is easier. Currently, it is using '_' as you indicated, but it could be modified.

**Comment by biancabnd** [comment]
Hi @RPS98, 

Looking at this PR, I still have the feeling that this implementation might not be suitable for everyone using the wrapper. Thus, I have the following suggestion for you:

-  change the `add_namespace_to_tf` parameter to a string parameter called `tf_frame_prefix` or similar
- append this parameter always to all frames in the wrapper
- for your specific case, you can then set this parameter in the your particular launch file to e.g. namespace + "/" . But this, will be your specific solution as not everyone might want to use the namespace for this. Also, this solves the issue between using "/" or using "_".  This also removes the need of you having to use different config files for each of your drones and makes the solution more generic and not limited to only using the namespace. 
- By default this parameter should be empty, so it will not affect the name of the frame id

**Comment by RPS98** [comment]
Hi @biancabnd,

Sounds good to me. I've already uploaded the changes.

P.S.: Part of the code that was there doesn't pass with either clang or uncrustify tests, it would need to be corrected.

**Comment by biancabnd** [comment]
> Hi @biancabnd,
> 
> Sounds good to me. I've already uploaded the changes.
> 
> P.S.: Part of the code that was there doesn't pass with either clang or uncrustify tests, it would need to be corrected.

Hi again, 
Thank you. Just merge the latest main in this branch and we will get this merged. The latest main includes some formatting checks. 

Thanks again for your time and collaboration! :confetti_ball:

**Comment by RPS98** [comment]
Hi @biancabnd,

Already done!

Thank you for your help. I hope to be able to use the main version of this repository in our system ([Aerostack2](https://aerostack2.github.io/)) without needing any forks in the near future!

**Review by biancabnd — APPROVED** [review]

**Pull request merged**